keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1316 @@
+   request and should wait for either the pod deletion or removal of the evacuation request. The
+   owning controller of the pod should observe the pod's conditions and respond to the
+   `EvacuationRequest` by accepting it and setting an `EvacuationInitiated` condition to `True` in
+   the pod conditions.
+2. `EvacuationInitiated` condition should be set by the owning controller to signal that work is
+   being done to either remove or evacuate/migrate the pod to another node. The draining
+   process/controller should wait a reasonable amount of time (3 minutes) to observe the appearance
+   of the condition or change of the condition status to `True`. The draining process should then
+   skip such a pod and leave its management to the owning controller. If `EvacuationInitiated`
+   condition does not appear after 3 minutes, the draining process will begin evicting or deleting
+   the pod. If the owning controller is unable to remove or migrate the pod, it should set the
+   `EvacuationInitiated` condition status back to `False` to give the eviction a chance to start.
+
+```golang
+
+type PodConditionType string
+
+const (
+    ...
+    EvacuationRequest PodConditionType = "EvacuationRequest"
+    EvacuationInitiated PodConditionType = "EvacuationInitiated"
+)
+
+const (
+    ...
+    PodReasonNodeMaintenance = "NodeMaintenance"
+)
+
+```
+### NodeMaintenance Controller
+
+Node maintenance controller will be introduced and added to `kube-controller-manager`. It will
+observe NodeMaintenance objects and have the following main features:
+
+#### Idle
+
+The controller should not touch the pods or nodes in any way in the `Idle` stage.
+
+### Finalizers and Deletion
+
+When a stage is not `Idle`, `nodemaintenance.kubernetes.io/maintenance-completion` finalizer will be put on
+the NodeMaintenance object to ensure uncordon and reset of EvacuationRequest conditions upon deletion.
+
+When a deletion of the NodeMaintenance object is detected, its `.spec.stage` is set to `Complete`.
+The finalizer is not removed until the `Complete` stage has been completed.
+
+#### Cordon
+
+When a `Cordon` or `Drain` stage is detected on the NodeMaintenance object, the controller
+will set `.spec.Unschedulable` to `true` on all nodes that satisfy `.spec.nodeSelector`.
+
+
+#### Uncordon (Complete)
+
+When a `Complete` stage is detected on the NodeMaintenance object, the controller will:
+- Set `.spec.Unschedulable` back to `false`  on all nodes that satisfy `.spec.nodeSelector`, unless
+  there is no other maintenance in progress.
+- Reset EvacuationRequest conditions on targeted pods (see Drain below for more details).
+
+
+#### Drain
+
+When a `Drain` stage is detected on the NodeMaintenance object, the
+`EvacuationRequest` condition is set on selected pods. The condition should have
+`Reason="NodeMaintenance$NAME"` and `Message="$REASON""`, where `NAME` is the name of the
+NodeMaintenance object and `REASON` is equal to `.spec.reason`.
```

Comment:
```suggestion
`Reason="NodeMaintenance` and `Message="$REASON""`, where `REASON` is equal to `.spec.reason`.
```

the `reason` field of conditions is supposed to be opaquely machine-readable; let's keep that.

- For example, one day we might support a way to query Pods based whether they have a condition with a particular reason.
- if someone creates a NodeMaintenance named (eg) `Cancelled` this could confuse people/

Reply:
I agree. I have been discussing this already in https://github.com/kubernetes/enhancements/pull/4479#issuecomment-1933246115 and need to figure out something better.

Reply:
All these evacuation issues should be solved by the new Evacuation KEP https://github.com/kubernetes/enhancements/pull/4565